compactor/src/handler.rs

```diff
@@ -414,6 +406,167 @@ async fn compact_hot_partitions(compactor: Arc<Compactor>) -> usize {
+            )
+            .await;
+
+        if let Err(e) = parquet_files_for_compaction {
+            warn!(
+                ?e,
+                ?partition_id,
+                "hot compaction failed due to error in reading parquet files"
+            );
+            // This may just be a hickup reading object store, skip commpacting it in this cycle
+            continue;
+        }
+        let parquet_files_for_compaction = parquet_files_for_compaction.unwrap();
+
+        // Return only files under the remaining_budget_bytes that should be compacted
+        let to_compact = parquet_file_filtering::filter_hot_parquet_files(
+            partition.clone(),
+            parquet_files_for_compaction,
+            remaining_budget_bytes,
+            columns,
+            &compactor.parquet_file_candidate_gauge,
+            &compactor.parquet_file_candidate_bytes,
+        );
+
+        // --------------------------------------------------------------------
+        // 3. Check the compactable status and act provide the right action
+        match to_compact.filter_result() {
+            FilterResult::NothingToCompact => {
+                continue;
+            }
+            FilterResult::ErrorEstimatingBudget => {
+                warn!(
+                    ?partition_id,
+                    ?table_id,
+                    "hot compaction is skipped due to error in estimating compacting memory"
+                );
+                // todo: add this partition and its info into a new catalog table
+                // https://github.com/influxdata/influxdb_iox/issues/5458
+                continue;
+            }
+            FilterResult::OverBudget => {
+                if remaining_budget_bytes < compactor.config.memory_budget_bytes {
+                    // Add this partition back the end of the list to compact it with full budget later
+                    candidates.push(partition);
+                } else {
+                    // Even with max budget, we cannot compact a bit of this partition, log it
+                    warn!(
+                        ?partition_id,
+                        ?table_id,
+                        "hot compaction is skipped due to over memory budget"
+                    );
+                    // todo: add this partition and its info into a new catalog table
+                    // https://github.com/influxdata/influxdb_iox/issues/5458
+                }
+                continue;
+            }
+            FilterResult::Proceeed => {
+                remaining_budget_bytes -= to_compact.budget_bytes();
+                parallel_compacting_candidates.push(to_compact);
+            }
+        }
+
+        // --------------------------------------------------------------------
+        // 4. Almost hitting max budget or no more candidates, let compact parallel_compacting_candidates
+        if (remaining_budget_bytes < (compactor.config.memory_budget_bytes as f64 * 0.5) as u64)
```

Comment:
Floating point can be expensive... because this is just trying to get 50%, can this be:

```suggestion
        if (remaining_budget_bytes < (compactor.config.memory_budget_bytes / 2) as u64)
```

and if we truncate off a byte, that should be ok?

Reply:
Yeah, they are the same